User profile with logout

zackads opened the pull request:
## Context

<!-- Why are you making this change? What might surprise someone about it? -->

Plays back the logged in user and allows user to log out by accessing the ActiveDirectory tenant:

![image](https://user-images.githubusercontent.com/54271433/118806896-883de280-b89f-11eb-92c6-3c65d508b0fc.png)

![image](https://user-images.githubusercontent.com/54271433/118806925-912eb400-b89f-11eb-93d1-4a6e03d12c79.png)

(Menu appears above the button that opens it.  Believe it or not, this is the [Material-UI pattern](https://material-ui.com/components/menus/)!  I haven't polished further as this provides the functionality, but can do on future iterations if necessary.)

Completes the task on the Registry/MCC/SAR team login [ticket](https://trello.com/c/7TxAPwUZ):

![image](https://user-images.githubusercontent.com/54271433/118806874-807e3e00-b89f-11eb-8a03-feea8957d447.png)

![image](https://user-images.githubusercontent.com/54271433/118808586-955bd100-b8a1-11eb-9058-d7b7d17ce426.png)

## Changes in this pull request

<!-- List all the changes, if there are UI changes, please include Before and After screenshots.  -->

- Create a [React context](https://reactjs.org/docs/context.html) to hold auth data about the current user.
- Implement UI to playback current auth context to user and allow user to log in / out.

## Guidance to review

<!-- How could someone else check this work? Which parts do you want more feedback on? -->

We could use the `useMsal()` [hook provided by the authentication library](https://reactjs.org/docs/context.html).  I spiked this initially and it works, but it would require every part of our application that touched auth/permissions to depend on the MSAL library.  So the `UserProfile` component would make calls directly on the `instance` provided by `useMsal()`.

This rang alarm bells for me as it would mean a high-level 'thing' in our app depending on a low-level implementation detail.  Even worse if we then let MSAL into our domain code. 🐉

To get around this, I created the shim component in `AuthWrapper` to abstract away the concrete MSAL property names behind an interface we control:

```typescript
  isAuthenticated: boolean;
  user: {
    username: string;
    displayName: string;
  };
  login: () => void;
  logout: () => void;
```

As we move on with role-based access control, we can add properties to this interface that expose things we need to pass to the API about the current user, IDs and claims etc.

Hope this makes sense and I haven't over-engineered?  Very grateful for feedback 🙏🏻

## Link to Trello card

<!-- https://trello.com/b/p2XQo8jN/beacons-beta-private -->

https://trello.com/c/7TxAPwUZ

## Things to check

- [-] Environment variables have been updated

Ninamma commented on src/components/auth/AuthWrapper.tsx:
```diff
@@ -1,23 +1,67 @@
+   * authenticated user data without depending on a concrete auth provider.
+   *
+   */
+  const currentUser = useMsal().instance.getAllAccounts()[0] || {};
```
Not 💯  but thinking that, you could just use pca here like this:
```suggestion
  const currentUser = pca.getAllAccounts()[0] || {};
```

beacuse we did something similar in the authGateway to get the current user: https://github.com/mcagov/beacons-backoffice/blob/cf9896bd11432ae8c7dddff115001f50b5bb3884/src/gateways/AuthGateway.tsx#L14

zackads replied:
Great thinking!  I had it this way initially, but the React component this code appears doesn't update the context without use of the `useMsal()` hook.  So it was working like this:

1.  Unauthenticated user first goes to webpage.  PCA is instantiated.
2. `<AuthWrapper>` renders and provides auth data to child components via React.context.  Because user is not yet authenticated, auth 'stuff' is empty.
3.  User authenticates.
4. Context provided by `<AuthWrapper>` is still empty because nothing has prompted it to re-render and update its values from `pca`.
5. So the user would log in but the `<UserMenu>` component would still think there was no authenticated user.

Using the `useMsal()` hook ([docs](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-react/docs/hooks.md#usemsal-hook)) means the shim'd context `<AuthContext>` provides downwards is updated from the MSAL context that we don't want to expose to our application's internals.  So now:

1 - 3 as above.
4. Context provided by `useMsal()` is updated, which prompts an update to our shim context provided by `<AuthWrapper>`, which becomes available to context consumers like `<UserMenu>`.
5. `<UserMenu>` can play back the authenticated user.

So whenever the user's authentication state changes `<AuthWrapper>` provides the most up-to-date context downwards.  I hope that makes sense, it took me a while to work out why using just `pca` wasn't updating after the user authenticated... 😬

Ninamma replied:
Ohh okay, that helps a lot, thank you for explaining it so well!! I can definitely imagine that not being very easy to figure out 😅

Ninamma reviewed (approved):
😍 looks great to me! so cool and a bit magic ✨

zackads commented:
@matthew-a-carr @Ninamma I've made two medium changes since last reviewed:

- Add unit tests on the `<UserMenu>`.
- Remove un-used attributes on `IAuthContext`: `isAuthenticated` and `login` are gone.  I put them on, but as Matt pointed out yesterday, YAGNI!

Just thought I'd check if you wanted to review again?
